The ``lazy_matrix_mul`` module
======================

Using ``lazy_matrix_mul``
-------------------

Importing the function from the module:
	>>> lazy_matrix_mul = __import__("101-lazy_matrix_mul").lazy_matrix_mul

Checking for module docstring:
	>>> m = __import__("101-lazy_matrix_mul").__doc__
	>>> len(m) > 1
	True

Checking for function docstring:
	>>> f = __import__("101-lazy_matrix_mul").lazy_matrix_mul.__doc__
	>>> len(f) > 1
	True

Checking for PEP8:
	>>> import os
	>>> os.popen("pep8 101-lazy_matrix_mul.py").read()
	''

Checking for last blank line:
	>>> os.popen("cat -e 101-lazy_matrix_mul.py | tail -1").read()[-1]
	'\n'

Checking for passing non-list as m_a:
	 >>> lazy_matrix_mul(None, [[1, 2], [3, 4]])
	 Traceback (most recent call last):
	 ...
	 TypeError: Object arrays are not currently supported

Checking for passing non-list to m_b:
	 >>> lazy_matrix_mul([[1, 2], [3, 4]], None)
	 Traceback (most recent call last):
	 ...
	 TypeError: Object arrays are not currently supported

Checking for passing empty m_a:
	 >>> lazy_matrix_mul([], [[1, 2], [3, 4]])
	 Traceback (most recent call last):
	 ...
	 ValueError: shapes (0,) and (2,2) not aligned: 0 (dim 0) != 2 (dim 0)

Checking for passing m_a with empty rows:
	 >>> lazy_matrix_mul([[], []], [[1, 2], [3, 4]])
	 Traceback (most recent call last):
	 ...
	 ValueError: shapes (2,0) and (2,2) not aligned: 0 (dim 1) != 2 (dim 0)

Checking for passing empty m_b:
	 >>> lazy_matrix_mul([[1, 2], [3, 4]], [])
	 Traceback (most recent call last):
	 ...
	 ValueError: shapes (2,2) and (0,) not aligned: 2 (dim 1) != 0 (dim 0)

Checking for passing m_b with empty rows:
	 >>> lazy_matrix_mul([[1, 2], [3, 4]], [[]])
	 Traceback (most recent call last):
	 ...
	 ValueError: shapes (2,2) and (1,0) not aligned: 2 (dim 1) != 1 (dim 0)

Checking for uneven m_a:
	 >>> lazy_matrix_mul([[1, 2], [3, 4, 5]], [[1, 2], [3, 4]])
	 Traceback (most recent call last):
	 ...
	 ValueError: setting an array element with a sequence.

Checking for uneven m_b:
	 >>> lazy_matrix_mul([[1, 2], [3, 4]], [[1, 2], [3, 4, 5]])
	 Traceback (most recent call last):
	 ...
	 ValueError: setting an array element with a sequence.

Checking for non-number in list in m_a:
	 >>> lazy_matrix_mul([[1, "Hello"], [3, 4]], [[1, 2], [3, 4]])
	 Traceback (most recent call last):
	 ...
	 TypeError: invalid data type for einsum

Checking for non-number in list in m_b:
	 >>> lazy_matrix_mul([[1, 2], [3, 4]], [[1, "Hello"], [3, 4]])
	 Traceback (most recent call last):
	 ...
	 TypeError: invalid data type for einsum

Checking for mismatching matrices:
	 >>> lazy_matrix_mul([[1, 2, 3], [4, 5, 6]], [[1, 2], [3, 4,]])
	 Traceback (most recent call last):
	 ...
	 ValueError: shapes (2,3) and (2,2) not aligned: 3 (dim 1) != 2 (dim 0)